scikeras/utils/__init__.py

```diff
@@ -0,0 +1,89 @@
+          ``continuous``.
+        * ``multilabel-indicator`` is more specific but compatible with
+          ``multiclass-multioutput``.
+        * ``multiclass-one-hot`` is more specific but compatible with
+          ``multilabel-indicator``.
+
+    Parameters
+    ----------
+    y : array-like
+
+    Returns
+    -------
+    target_type : string
+        One of:
+
+        * 'continuous': `y` is an array-like of floats that are not all
+          integers, and is 1d or a column vector.
+        * 'continuous-multioutput': `y` is a 2d array of floats that are
+          not all integers, and both dimensions are of size > 1.
+        * 'binary': `y` contains <= 2 discrete values and is 1d or a column
+          vector.
+        * 'multiclass': `y` contains more than two discrete values, is not a
+          sequence of sequences, and is 1d or a column vector.
+        * 'multiclass-multioutput': `y` is a 2d array that contains more
+          than two discrete values, is not a sequence of sequences, and both
+          dimensions are of size > 1.
+        * 'multilabel-indicator': `y` is a label indicator matrix, an array
+          of two dimensions with at least two columns, and at most 2 unique
+          values.
+        * 'multiclass-one-hot': `y` is a label indicator matrix, an array
+          of two dimensions with at least two columns, and at most 2 unique
+          values where each sample has only 1 class assigned.
+        * 'unknown': `y` is array-like but none of the above, such as a 3d
+          array, sequence of sequences, or an array of non-sequence objects.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> type_of_target([0.1, 0.6])
+    'continuous'
+    >>> type_of_target([1, -1, -1, 1])
+    'binary'
+    >>> type_of_target(['a', 'b', 'a'])
+    'binary'
+    >>> type_of_target([1.0, 2.0])
+    'binary'
+    >>> type_of_target([1, 0, 2])
+    'multiclass'
+    >>> type_of_target([1.0, 0.0, 3.0])
+    'multiclass'
+    >>> type_of_target(['a', 'b', 'c'])
+    'multiclass'
+    >>> type_of_target(np.array([[1, 2], [3, 1]]))
+    'multiclass-multioutput'
+    >>> type_of_target([[1, 2]])
+    'multilabel-indicator'
+    >>> type_of_target(np.array([[1.5, 2.0], [3.0, 1.6]]))
+    'continuous-multioutput'
+    >>> type_of_target(np.array([[0, 1], [1, 1]]))
+    'multilabel-indicator'
+    >>> type_of_target(np.array([[0, 1], [1, 0]]))
+    'multiclass-one-hot'
+    """
+    type_ = _type_of_target(y)
+    if type_ == "multilabel-indicator":
+        if len(y.shape) == 2 and np.all(np.sum(y, axis=1) == 1):
+            # one-hot encoded target
+            # this is a subset of multilabel-indicator
+            # but SciKeras differentiates it since it is
+            # a common use case in Keras to have a single multiclass
+            # target one-hot encoded + categorical_crossentropy as a loss
+            type_ = "multiclass-one-hot"
```

Comment:
In Keras it's pretty common to one-hot-encode your target and use that with categorical cross-entropy as the loss. That maps 1:1 with a use case in Keras. On the other hand, any other `"multilabel-indicator"` does not map directly to a use case in Keras. This tries to differentiate the two to avoid confusing users with some cryptic TF error if they pass a `"multilabel-indicator"` target.

Reply:
When does a "cryptic TF error" get raised, and what's the error message look like? Does categorical cross-entropy loss work with multiple class labels per row? Say a "two-hot encoding" where every example has two class labels. 

What else does this do besides potentially avoid user confusion? To me, "multiclass indicator" is more clear than "multiclass onehot" because "indicator" has a clear definition and "onehot" is jargon.

I don't like unnecessarily modifying the Scikit-learn API, especially because an invalid target type can raise an error. In my view, SciKeras is a pipe between Keras and Scikit-Learn. I'm not seeing why this is a required necessary modification.

Reply:
So you think that for `multilabel-indicator` and `multiclass-multioutput` we should pass throu without an error or warning?

Reply:
An example error scenario. Target is 1 binary + 1 multiclass == multiclass-multioutput:

```python3
import numpy as np
from sklearn.datasets import make_classification
from tensorflow import keras


X, y_bin = make_classification(n_classes=2)
_, y_cat = make_classification(n_classes=3, n_informative=3)
y = np.column_stack([y_bin, y_cat])

inp = keras.layers.Input(shape=(X.shape[1], ))
x1 = keras.layers.Dense(100)(inp)  # hidden layer
binary_out = keras.layers.Dense(1, activation="sigmoid")(x1)
cat_out = keras.layers.Dense(3, activation="softmax")(x1)
losses = ["binary_crossentropy", "sparse_categorical_crossentropy"]
model = keras.Model([inp], [binary_out, cat_out])
model.compile(loss=losses)

# assuming user overrode preprocess_y and split y into a list
model.fit(X, [y_bin, y_cat])  # works fine
# assuming SciKeras did nothing, passing through y to Keras
model.fit(X, y)  #  ValueError
```

This raises `ValueError: logits and labels must have the same shape ((None, 1) vs (None, 2))` within `tensorflow/python/ops/nn_impl.py`. That's not very user friendly, and doesn't really pinpoint where you would implement a fix. By raising a ValueError in `preprocess_y` we can explicitly say "hey, override this function to get your expected behavior, and here are links to examples on how to do it"

Reply:
> Does categorical cross-entropy loss work with multiple class labels per row? Say a "two-hot encoding" where every example has two class labels.

It might run (as in, no error) but that's not how you should solve that problem (or at least not how I would solve that problem). I would set that up as 2 outputs with sigmoid activation and `binary_crossentropy` as the loss function (the actual loss function becomes the average of that loss of each output).

Reply:
Let's move followup discussion to https://github.com/adriangb/scikeras/pull/88#issuecomment-703828162 to keep this important issue at the top-level.